Fix descriptor lock release logic for block volume unmapDevice

mkimuram opened the pull request:
**What this PR does / why we need it**:
Fix descriptor lock release logic for block volume ```unmapDevice```

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #69114

**Special notes for your reviewer**:
/sig storage

**Release note**:
```release-note
NONE
```

mkimuram committed:
Fix descriptor lock release logic for block volume unmapDevice

Fixes: #69114

mkimuram commented:
@rootfs @vladimirvivien 

I opened separate issue and create separate PR for descriptor lock issue. PTAL

Noth that I tested e2e volumeMode test with below options and confirmed that it succeeds.

  - --ginkgo.focus="In-tree.*iscsi.*Pre-provision.*block.*volumeMode.*should create.*"
  - --ginkgo.focus="In-tree.*rbd.*Pre-provision.*block.*volumeMode.*should create.*"


Also, while executing above tests, below command shows that descriptor lock is created and deleted properly on pod creation and deletion.

```
# while :;do date;losetup;sleep 1;done
<snip>

Wed Sep 26 19:27:33 UTC 2018
Wed Sep 26 19:27:34 UTC 2018
Wed Sep 26 19:27:35 UTC 2018
NAME       SIZELIMIT OFFSET AUTOCLEAR RO BACK-FILE DIO
/dev/loop0         0      0         0  0 /dev/sda    0
Wed Sep 26 19:27:36 UTC 2018
NAME       SIZELIMIT OFFSET AUTOCLEAR RO BACK-FILE DIO
/dev/loop0         0      0         0  0 /dev/sda    0

<snip>

Wed Sep 26 19:27:43 UTC 2018
NAME       SIZELIMIT OFFSET AUTOCLEAR RO BACK-FILE DIO
/dev/loop0         0      0         0  0 /dev/sda    0
Wed Sep 26 19:27:44 UTC 2018
Wed Sep 26 19:27:45 UTC 2018

<snip>

Wed Sep 26 19:32:37 UTC 2018
Wed Sep 26 19:32:38 UTC 2018
NAME       SIZELIMIT OFFSET AUTOCLEAR RO BACK-FILE DIO
/dev/loop0         0      0         0  0 /dev/rbd0   0
Wed Sep 26 19:32:39 UTC 2018
NAME       SIZELIMIT OFFSET AUTOCLEAR RO BACK-FILE DIO
/dev/loop0         0      0         0  0 /dev/rbd0   0
Wed Sep 26 19:32:40 UTC 2018

<snip>

Wed Sep 26 19:32:44 UTC 2018
NAME       SIZELIMIT OFFSET AUTOCLEAR RO BACK-FILE DIO
/dev/loop0         0      0         0  0 /dev/rbd0   0
Wed Sep 26 19:32:45 UTC 2018
Wed Sep 26 19:32:46 UTC 2018
```

rootfs commented:
/approve

rootfs commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -1113,6 +1099,21 @@ func (og *operationGenerator) GenerateUnmapDeviceFunc(
+		removeMapPathErr := og.blkUtil.RemoveMapPath(globalMapPath)
+		if removeMapPathErr != nil {
+			// On failure, return error. Caller will log and retry.
+			return deviceToDetach.GenerateError("UnmapDevice failed", removeMapPathErr)
```
keep original msg `UnmapDevice.RemoveLoopDevice failed`

mkimuram replied:
@rootfs 

I changed from "UnmapDevice" to "UnmapDevice.RemoveMapPath" as suggested.
Also, I rebased the commit. PTAL

rootfs commented:
this looks good to me, @msau42 can you take a look?

msau42 commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -1079,30 +1079,16 @@ func (og *operationGenerator) GenerateUnmapDeviceFunc(
 			} else {
-				errInfo := "UnmapDevice.GetLoopDevice failed to get loopback device, " + fmt.Sprintf("device path: %q", deviceToDetach.DevicePath)
-				return deviceToDetach.GenerateError(errInfo, err)
+				glog.Warningf(deviceToDetach.GenerateMsgDetailed("UnmapDevice: GetLoopDevice failed to get loopback device",
```
what other errors would make it ok to skip the error and continue?

mkimuram replied:
@msau42

Thank you for pointing it out.

I thought that I hit an issue by not skiping the error from ```GetLoopDevice``` other than ```ErrDeviceNotFound``` error, while I tested. However, I can't reproduce the issue, now.

```unmapDeviceFunc``` that is generated in ```GenerateUnmapDeviceFunc``` could be called multiple times, so there will be cases that ```GetLoopDevice``` is called after ```RemoveLoopDevice``` or ```TearDownDevice``` from previous ```unmapDeviceFunc``` calls. 
As you pointed out, in most cases, ```GetLoopDevice``` returns ```ErrDeviceNotFound``` error, so the only cases that I can think of are:
1. ```os.Stat({path})``` fails with error other than not exist error,
2. ```losetup -f {path}``` command fails with error.

I guess that the second one would much likely to happen because man page of losetup is saying that it could return erorr, but I can't make losetup return these errors.
```
RETURN VALUE
       losetup returns 0 on success, nonzero on failure.  When losetup displays the status of a loop device, it returns 1 if the device is not con‐
       figured and 2 if an error occurred which prevented determining the status of the device.
```

So, I started to think that it would be better not changing here, until we find a good reproducer. I will revert the change for skipping error.

msau42 commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -1113,6 +1099,21 @@ func (og *operationGenerator) GenerateUnmapDeviceFunc(
 			}
 		}
 
+		// Execute tear down device
```
Can you add a comment explaining that we need to call TearDownDevice afterwards because plugins that do local attach/detach will fail due to the refcnt on the loopback device.

mkimuram replied:
I will add a comment.

mkimuram committed:
Fix to reflect commnet

- Change not to skip error from GetLoopDevice other than DeviceNotFound
- Add comment for the reason for order of descriptor lock release and TearDownDevice

mkimuram commented:
@msau42 

I fixed as suggested. PTAL

msau42 commented:
/ok-to-test

msau42 commented:
/lgtm

mkimuram commented:
/kind bug
